lib/puppet/functions/stdlib/getpwnam.rb

```diff
@@ -0,0 +1,27 @@
+# frozen_string_literal: true
+
+# @summary
+#   Given a username returns the user's entry from the `/etc/passwd` file.
+#
+#
+Puppet::Functions.create_function(:'stdlib::getpwnam') do
+  # @param user
+  #    The username to fetch password record for.
+  #
+  # @example Get a password entry for user steve as a hash.
+  #   $passwd_entry = stdlib::getpwnam('steve')
+  #
+  # @example Get the UID of user steve
+  #   $uid = stdlib::getpwnam('steve')["uid"]
+  #
+  # @return
+  #   [Hash] For example {"name"=>"root", "passwd"=>"x", "uid"=>0, "gid"=>0, "gecos"=>"root", "dir"=>"/root", "shell"=>"/bin/bash"}
+  dispatch :getpwnam do
+    param 'String', :user
+    return_type 'Hash'
+  end
+
+  def getpwnam(user)
+    Etc.getpwnam(user).to_h.first(7).map { |k, v| { k.to_s => v } }.reduce(:merge)
```

Comment:
If the user isn't found this will probably end up being `nil.to_h`. Also, why do you even need to do the map? Isn't this sufficient?
```suggestion
    Etc.getpwnam(user)&.to_h
```

I'd expect Puppet to translate the symbols to strings.

Reply:
Without just `Etc.getpwnam(user).to_h` you get a Tuple rather than a Hash back - changing the return to Tuple 
shows this:
```
       expected stdlib::getpwnam("steve") to have returned {"name"=>"steve", "passwd"=>"x", "uid"=>1000, "gid"=>1001, "gecos"=>"Steve", "dir"=>"/home/steve", "shell"=>"/bin/fish"} instead of [[:name, "steve"], [:passwd, "x"], [:uid, 1000], [:gid, 1001], [:gecos, "Steve"], [:dir, "/home/steve"], [:shell, "/bin/fish"]]
```
which does not seem much good.

For the no such user it fails in Etc.getpwnam itself.

```
ArgumentError(can't find user for notexisting)
```
which seemed reasonable to me.

Reply:
> With just `Etc.getpwnam(user).to_h` you get a Tuple rather than a Hash back - changing the return to Tuple

This is odd, perhaps Ruby version dependent?

```console
$ ruby --version
ruby 3.2.2 (2023-03-30 revision e51014f9c0) [x86_64-linux]
$ ruby -retc -e "puts Etc.getpwnam('ekohl').to_h"
{:name=>"ekohl", :passwd=>"x", :uid=>1000, :gid=>1000, :gecos=>"ekohl", :dir=>"/home/ekohl", :shell=>"/bin/bash"}
```

> which seemed reasonable to me.

Sounds fair to me. Is there a `@raises`  tag in puppet-strings to document this?

Reply:
Added the `@raises`

With `Etc.getpwnam(user).to_h` :

* ruby 3.2.2 and puppet 8.5.0    (should switch to 3.3.X I think)
* ruby 2.7.7 and puppet 7.29.0 

I get 
```
[[:name, "steve"], [:passwd, "x"], [:uid, 1000], [:gid, 1001], [:gecos, "Steve"], [:dir, "/home/steve"], [:shell, "/bin/fish"]]
````
for both.

Reply:
Hum, I checked on a bunch of versions of ruby and `Etc.getpwnam('romain').to_h` return a Hash for all of them (2.5, 2.7, 3.1 and 3.2).  What you get in your last message seems to be the output of `to_a` applied on a Hash :thinking:.

Also why only keep the 7 first entries? (also `#first` on a Hash return an Array :wink:) All OS do not return the same information, so I think it make sense to keep it unchanged and allow the end-user to consume what they want from the hash?

The suggested change by @ekohl seems to really be what I would expect, can you please double-check this and if you reproduce the issue show us the failing code so that we can better understand what is going on?